db/db.go

```diff
@@ -168,154 +169,195 @@ func newDB(n app.Net, id thread.ID, options *NewDBOptions) (*DB, error) {
 			return err
 		}
-
-		var indexes map[string]IndexConfig
+		c, err := newCollection(name, schema, d)
+		if err != nil {
+			return err
+		}
+		var indexes map[string]Index
 		index, err := d.datastore.Get(dsDBIndexes.ChildString(name))
 		if err == nil && index != nil {
-			_ = json.Unmarshal(index, &indexes)
-		}
-
-		indexValues := make([]IndexConfig, len(indexes))
-		for _, value := range indexes {
-			indexValues = append(indexValues, value)
+			if err := json.Unmarshal(index, &indexes); err != nil {
+				return err
+			}
 		}
-
-		if _, err := d.NewCollection(CollectionConfig{
-			Name:    name,
-			Schema:  schema,
-			Indexes: indexValues,
-		}); err != nil {
-			return err
+		for _, index := range indexes {
+			if index.Path != "" { // Catch bad indexes from an old bug
+				c.indexes[index.Path] = index
+			}
 		}
+		d.collections[c.name] = c
 	}
 	return nil
 }
 
+// GetDBInfo returns the addresses and key that can be used to join the DB thread.
+func (d *DB) GetDBInfo(opts ...Option) ([]ma.Multiaddr, thread.Key, error) {
+	options := &Options{}
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	tinfo, err := d.connector.Net.GetThread(context.Background(), d.connector.ThreadID(), net.WithThreadToken(options.Token))
+	if err != nil {
+		return nil, thread.Key{}, err
+	}
+	return tinfo.Addrs, tinfo.Key, nil
+}
+
 // CollectionConfig describes a new Collection.
 type CollectionConfig struct {
 	Name    string
 	Schema  *jsonschema.Schema
-	Indexes []IndexConfig
+	Indexes []Index
 }
 
-// NewCollection creates a new collection in the db with a JSON schema.
-func (d *DB) NewCollection(config CollectionConfig) (*Collection, error) {
+// NewCollection creates a new db collection with config.
+// @todo: Handle token auth
+func (d *DB) NewCollection(config CollectionConfig, opts ...Option) (*Collection, error) {
 	d.lock.Lock()
 	defer d.lock.Unlock()
 
-	if _, ok := d.collectionNames[config.Name]; ok {
-		return nil, fmt.Errorf("already registered collection")
+	if _, ok := d.collections[config.Name]; ok {
+		return nil, ErrCollectionAlreadyRegistered
 	}
-
 	c, err := newCollection(config.Name, config.Schema, d)
 	if err != nil {
 		return nil, err
 	}
-	key := dsDBSchemas.ChildString(config.Name)
-	exists, err := d.datastore.Has(key)
-	if err != nil {
+	if err := d.addIndexes(c, config.Schema, config.Indexes, opts...); err != nil {
 		return nil, err
 	}
-	if !exists {
-		schemaBytes, err := json.Marshal(config.Schema)
-		if err != nil {
-			return nil, err
-		}
-		if err := d.datastore.Put(key, schemaBytes); err != nil {
-			return nil, err
-		}
+	if err := d.saveCollection(c); err != nil {
+		return nil, err
 	}
+	return c, nil
+}
+
+// UpdateCollection updates an existing db collection with a new config.
+// Indexes to new paths will be created.
+// Indexes to removed paths will be dropped.
+// @todo: Handle token auth
+func (d *DB) UpdateCollection(config CollectionConfig, opts ...Option) (*Collection, error) {
```

Comment:
Nit: possible improvement for the future, validate all actions, then do actions. Just to minimize the possibility of partial applied changes. Not sure if it may worth it since the client could retry knowing with the "remaining" part (which might be a bit hard to know which one, but anyway).

Reply:
Nice yeah, I had a similar thought. I'll create an issue.